### ISSUE TITLE
remove TensorNetwork attribute from Node

### ISSUE THREAD

**Issue by mganahl** [google/TensorNetwork#259]
This is just raising some of the observations in #191 into a new issue again
 `TensorNetwork` object in `Node` is primarily used for backend access. 
I think we could  replace the network attribute with a backend attribute, and redesign
`TensorNetwork` such that current checks performed in `Node` by accessing `TensorNetwork` can be done directly from the network (or we even drop them).
This would also fit nicely with a standalone `Node` object.

**Comment by Thenerdstation** [comment]
I'm still not convinced what we gain by having nodes not associated with a TensorNetwork. I know what we lose, we can no longer do `contractors.optimal(net)` easily.

Can you give an example of what's easier with free nodes?

**Comment by mganahl** [comment]
The main thing is that the API becomes simpler and more intuitive without losing any of its flexibility. Conceptually it makes much more sense to me to switch the hierarchy between `Nodes` and `TensorNetwork`.  As @viathor has pointed out, there are many different networks that can be built from a single set of nodes. I think it would be very natural to have `Nodes` and `Edges` as fundamental objects on which operations are performed directly. If we need a `TensorNetwork` object,  we can still define it by first choosing a subset of the `Nodes`, then connecting them, and then pass all these `Nodes` to a `TensorNetowrk` constructor.
This also mimics very nicely the way that we use tensornetworks in practice. Take for example MPS algorithms: most of them are in fact a collection of very many smaller subroutines, each of which operate only on a few nodes out of a large number of total nodes. So what I want is 
clean, simple and easy way to build these "subnetworks" and do stuff on them (which in 99% of the cases is contraction, transposition, conjugaiton, svd, qr, and eigh). In the majority of cases I don't care if `Nodes` or `Edges` have unique names, and if I can access them using a str or not (even though this would still all be possible with free `Nodes` as well).
We can easily add a function `contractors.optimal([a,b,c,...])` operating on`Nodes` instead of `TensorNetwork` that finds optimal contraction paths (likewise for other functions).
The main purpose of `TensorNetwork` right now is to make sure that `Nodes` and `edges` have unique names (and some tests that I think can be discarded). We can still create a `TensorNetwork`
from a set of connected `Node` objects: `TensorNetowrk([a,b,c,..])` at which point we can easily check if all `Node` and `Edge` objects have unique names, and if the network is fully connected or not. There is nothing lost, but what we gain is conceptual simplicity.
 
a concrete example: again MPS (which is what the majority of people care about). First, I need to create my `Nodes` by a call to `TensorNetwork`. Then in order to compte for example correlators, I have to constantly create new `TensorNetwork` objects, add two tensors from two MPS nodes to them, contract a small network of three nodes, and throw the tensornetwork object away again. It's really an overkill to create a `TensorNetwork` at each step in a for loop: I want to perform a single contraction, but I have to create this object with all its methods and attributes, just to throw it away the next instant.  For my taste, there are just way too many `TensorNetworks` hanging around (but that's of course my subjective opinion :))

**Comment by MichaelMarien** [comment]
> r my taste, there are just way too many `TensorNetworks` hanging around (but that's of course my subjective opinion :))

I agree with this. Was porting some old Matlab code a while back and notices the exact same issue. Especially for uniform MPS, it is extremely natural to define them on the level of nodes (the TensorNetwork is infinite and never constructed), so I basically have a complete overhead with a uMPS TensorNetwork of which is only access the single Node.

**Comment by mganahl** [comment]
Hey @MichaelMarien  @amilsted @Thenerdstation  feel free to check out this experimental code https://github.com/mganahl/TensorNetwork/tree/free_node_no_net
on my fork. This is an experimental branch where I am restructing the API there as described above. I'm finished with seperating `TensorNetwork` from `Node`, such that `Node` is now freely usable 
with `tn.connect`, `tn.contract`, `tn.contract_between, tn.conj, tn.split_node_..., tn.transpose`.
I would be very interested in your opinions. Right now I'm implementing the changes to `TensorNetwork` now. The plan is to have it initialized from a set of connected, uniquely named nodes and edges.

I haven't yet touched upon opt_einsum contractors, but this should be pretty straightforward once 
`TensorNetwork` is back online.

**Comment by Thenerdstation** [comment]
>  We can still create a TensorNetwork from a set of connected Node objects: TensorNetowrk([a,b,c,..])

This is a dealbreaker for me, as it is very tedious to have to keep track of all the nodes. If a user needs to store all of the nodes they have to use these contractors, they won't use the contractors. Also, `TensorNetwork` objects are very very cheap to build/use and they give us a ton of checks for common errors. I don't see how creating temporary ones is bad at all.

**Comment by Thenerdstation** [comment]
@mganahl I would really recommend against doing a full rewrite until we have made a decision as to whether we will do this or not. I'd hate for you to put in a ton of work just for the PR to be closed...

**Comment by mganahl** [comment]
Don't worry, it wasn't too much work. I really needed a working baseline to see how useful it is 😄, and I have to say I _really_  came to like the free `Node`.

The thing is with the current API we already have to store all the nodes explicitly (for MPS, PEPS, MERA, ...), and I don't see any way of not doing that. In fact, one way or another we either have to store all out needed tensor objects in a list and keep on feeding them into `TensorNetwork` to use an optimal contractor (meaning there will be lots of calls to `TensorNetwork.add_node`), or we decide to store just `Nodes` instead and pass a list.
I don't see why this 


```python
nodes = [a,b,c,d,e] #some nodes
#connect nodes:
nodes[0][0] ^ nodes[2][1]
...
tn.contractors.optimal(nodes) #or tn.contractors.optimal(TensorNetwork(nodes))
```
is so much more terrible than this:
```python 
tensors = [a,b,c,d,e] #some tensors
net = TensorNetwork()
nodes=[net.add_node(t) for t in tensors]
#connect nodes:
nodes[0][0] ^ nodes[2][1]
...
tn.contractors.optimal(net)
```


For most of the tensor network stuff out there, automatic contractors are not immediately important, so for me this would not be a dealbreaker at all.  For the cases where they are nice (MERA and PEPS for example) the relevant networks are mostly small, and making lists with nodes is not an issue there.

As for the checks, I think all important ones can actually be performed without the `TensorNetwork`.  For me the two most important features of `TensorNetwork` are 1. that a tensor axis can only connect to one other tensor axis and 2. that edges are passed on to the outcome of a contraction. Once this is established, you really need to actively try to break the API.
In fact, one of the most typical errors that appear in e.g. infinite MPS are flipped indices between the input to a sparse transfer matrix solver and the output, and this case is not detectable with our API. For example for infinite MPS, there are a lot of places where we have to do stuff like this:
```python
def left_tm(tensor,left_reduced):
  net=TensorNetwork()
  A = net.add_node(tensor)
  Abar = net.add_node(net.backend.conj(tensor))
  left = net.add_node(left_reduced)
  A[0] ^ left[0]
  Abar[0] ^ left[1]
  A[1] ^ Abar[1]
  result = l @ A @ Abar
  return result.tensor #oops forgot to ensure correct index ordering
```
Also, doesn't this look much nicer :) ?
```python
def left_tm(A,left): #A and l are nodes here
  A[0] ^ left[0]
  tn.conj(A)[0] ^ left[1]
  A[1] ^ tn.conj(A)[1]
  return l @ A @ Abar #oops forgot to ensure correct index ordering
```

**Comment by Thenerdstation** [comment]
Ok now I'm starting to see your point. These are much nicer.

I think the easiest thing to do to relieve my anxiety around the contractors is to add a `tn.reachable(node)` method that returns a set of the nodes that are reachable from the given node. Our contractors need connected graphs anyway.

If we do this, then we need to also nuke the `TensorNetwork` object entirely. (I always subscribe to only having one way to do things: the right way.) This is going to break all of our tests and will make testing all of the backends harder. However, if we're careful it shouldn't be too hard.

Is the plan then to just have all of nodes after contraction just have new dangling edges instead of being disabled?

**Comment by amilsted** [comment]
@mganahl This looks rather nice! A network is just a collection of `Node`s, so why have a special container class? Overhead from `TensorNetwork` objects in the python code doesn't seem very important to me (we usually want to "compile out" the python for speed anyway!), but this does seem like a big conceptual simplification: We no longer need to have `merge` or `add_subnetwork`, since these operations just become `connect()`s. We would indeed want a `tn.reachable()` (or maybe `tn.connected_nodes()` for clarity?), I think, but that's easy to do. 

Here are a some other questions/comments:

1. Node-passing style: How does the caller of functions like `left_tm()` handle the fact that contractions mutate the `Node`s? I see that contracting currently "resets" a `Node` by giving it new dangling edges, but what if the function doesn't contract the Node? I suppose if the caller wants to ensure any `Node`s passed maintain any existing connections, a "Node.copy()" implementation that produces a new, disconnected copy of the node will be needed?
2. Comment on 1: Right now, possible confusion from mutation of `Node`s seems like the main argument for having explicit network objects - they provide an explicit context for operations, reminding the user of dependencies (also why I argued for network context managers).
3. We haven't actually been checking for uniqueness of `Node` and `Edge` names in `TensorNetwork`, AFAICT. The lack of uniqueness is the reason we couldn't come up with nice rules for preserving axis names, no?
4. Why do you never use axis labels? ;)

**Comment by mganahl** [comment]
Hey @amilsted, yes, those functions would indeed become superfluous. I was actually in favor of @Thenerdstation's suggestion to remove `TensorNetwork` at all, but one reason to keep it could be to support persistent networks (as you raise in 1.). Instead of `Node.copy()`, I would prefer to have `Node.__init__(node)` to also take a `Node` object, in which case it just creates a new `Node` with dangling edges, but shares the data with `node` (to be not misleading, I think`Node.copy()` should really produce a copy of the object, including its edges and their connections). Thus having static networks could work like 
```python
#define a static tensor network in `nodes`
static_nodes = [Node(np.random(...)) for _ in range(N)]
edges = [static_nodes[n][1] ^ static_nodes[n+1][0] for n in range(N-1)]

def fun1(a,b,c,...):
  a[0] ^ b[0]
  c[0] ^ b[1]
  ...
#create new nodes to preserve connections
a,b,c = Node(static_static_nodes[0]), Node(static_nodes[1]), Node(static_nodes[2])
result = fun(a,b,c)
```
I don't see any problems it `fun` doesn't contract a node. Connections of the node are in this case unaltered, so any assumptions that the user held before the call to `fun` about the node are still
valid after. 

If we want to be more explicit than just a list of connected nodes, we can keep `TensorNetwork`, for example by initializing it from a set of connected nodes via a `Node.copy()`, even though I would actually prefer just having the list. 

You're actually right, I just checked and `TensorNetwork` is only ensuring uniqueness if no `name` arguments are passed to `add_node` (the docstring says otherwise, so I'm not sure if this shouldnt be marked as a bug). I also noticed that `def _new_node_name` and `def _new_edge_name` have `Optional` parameters with no defaults (I've noticed this also at a few other places, not sure why `pytype` is not catching this). 

And yes, I've never need axis labels so far, but it could potentially free you from index conventions....

**Comment by amilsted** [comment]
> I don't see any problems it fun doesn't contract a node. Connections of the node are in this case unaltered, so any assumptions that the user held before the call to fun about the node are still
valid after.

I should have been clearer: I was imagining a function that *does* change the connections of a `Node` but does not contract it (perhaps to produce a network to be passed to another function).

**Comment by amilsted** [comment]
> Instead of Node.copy(), I would prefer to have Node.__init__(node) to also take a Node object

Agreed, this is better.

**Comment by amilsted** [comment]
By the way: I agree with @Thenerdstation and @mganahl that we should either do free nodes or explicit networks, but probably not both.

**Comment by mganahl** [comment]
> > I don't see any problems it fun doesn't contract a node. Connections of the node are in this case unaltered, so any assumptions that the user held before the call to fun about the node are still
> > valid after.
> 
> I should have been clearer: I was imagining a function that _does_ change the connections of a `Node` but does not contract it (perhaps to produce a network to be passed to another function).

I think this is also fine. In this case the user knows that `fun` is altering connections. If 
`fun(a,b,c)` expects nodes that are connected in a specific way to alter them into a specific other network, then in its body it has to disconnect connected edges and reconnect unconnected ones. Apart from the fact that such a thing is inherently bug-prone, I can't see any problems: if the user accidentally passes `a,b,c` with unexpected connections, this will likely raise an exception (due to trying to disconnect/connect nodes that are not connected/disconnected).

**Comment by amilsted** [comment]
Perhaps a `tn.copy_subnet()` would be handy for picking out parts of a predefined network. I can imagine using this to avoid doing pointless computations in a gauged MPS, for example.
```python
mps = [Node(np.random(...), axis_names=["L", "P", "R"]) for _ in range(N)]
bonds = [mps[n]["R"] ^ mps[n+1]["L"] for n in range(N-1)]

def expect_2site_centergauge(mps, n, op):
  A1, A2 = tn.copy_subnet(mps[n:n+2])
  B1, B2 = tn.conj_subnet(mps[n:n+2])
  A1["L"] ^ B1["L"]
  A2["R"] ^ B2["R"]
  op["in_1"] ^ A1["P"]
  op["in_2"] ^ A2["P"]
  op["out_1"] ^ B1["P"]
  op["out_2"] ^ B2["P"]
  result = tn.contractors.optimal(tn.connected_nodes(A1))
  return result.tensor
```

**Comment by mganahl** [comment]
That would indeed be useful to have. `tn.copy_subnet(nodes)` would automatically set all connections that point outside `nodes` to `dangling`. Maybe call it `create_subnet`?

**Comment by mganahl** [comment]
I've added `tn.reachable` (pretty easy to do). If we decide to do this, I can open a PR to create a new branch `free_node` or something, where we can start to work in this collaboratively.

**Comment by viathor** [comment]
It looks like this proposal would turn code that uses `TensorNetwork` instances as a container for the nodes and edges into code that uses python list as a container for nodes (with edges being carried around by the nodes).

I like this since it replaces a class we have to maintain with a standard python container while simplifying a lot of code as in the snippets above. It also makes sense conceptually which translates directly into a practical benefit: creating *ad hoc* tensor networks becomes very simple and basically corresponds to putting a bunch of nodes together into a python list.

I have two simple comments, a technicality and a design principle I think we should adopt. First, the technicality: a tensor network is really a *set* of nodes, i.e. the order of the nodes is not generally relevant. That said, in practice I do expect people to often use python list. For this reason, I think the code should work with any collection type (i.e. anything you can iterate over to get nodes or against which you can perform node membership checks) including lists and sets. In mypy notation: `Collection[Node]`.

Second, the principle: I think we can make this even more elegant and practical by employing the following simple rule:

*Functions that work on a Node instance, should work on a collection of Nodes, too.*

In terms of API, this would mean that any function like

```
conj(node: Node) -> Node
```

becomes

```
conj(node: Union[Node, Collection[Node]]) -> Union[Node, Collection[Node]].
```

The rationale for this is that a collection of nodes can be thought of as a logical node. The edges of the logical node are all the edges of the nodes in the collection except the internal edges between the nodes. In fact, one could obtain a cold, hard Node instance corresponding to a collection of nodes by contracting all internal edges. Of course we don't generally want to do this for efficiency reasons. In fact, AFAIU the essential benefit of using tensor networks is that they enable efficient representation of high-order tensors using a collection of smaller low-order tensors. Therefore it is crucial that a tensor networks library should treat node collections as first class citizens and allow the users to manipulate node collections directly without writing out all collection traversals explicitly each time.

In practice, in order to maintain the illusion that a collection of nodes *is* a node, we'd need a few utility functions such as `edge(node: Union[Node, Collection[Node]], axis_name: str) -> Edge`. What this would do when passed in a collection of nodes is return the externally-facing edge of the input nodes. There is an issue here regarding unique identification of an axis of a node collection that I leave open for discussion.

In addition to conceptual sanity, adhering to this principle would allow even more flexibility in user code. It would also mean that certain tensors get a natural, efficient representation as a tensor network. For example, the quantum CZ gate becomes a 3-tuple of two copy tensors and a phasing tensor:

```python
h = np.array([[1, 1], [1, -1]])/np.sqrt(2)
phases = tn.Node(h, axis_names=['q0', 'q1'])
cn0 = tn.CopyNode(rank=3, dimension=2,
                  axis_names=['q0in', 'q0out', 'phase_pickup'])
cn1 = tn.CopyNode(rank=3, dimension=2,
                  axis_names=['q1in', 'q1out', 'phase_pickup'])
edge(cn0, 'phase_pickup']) ^ edge(phases, 'q0')
edge(cn1, 'phase_pickup']) ^ edge(phases, 'q1')
CZ = [cn0, phases, cn1]
assert edges(CZ), {'q0in', 'q1in', 'q0out', 'q1out'}
```

Now, despite CZ being a tuple of nodes rather than a node instance, you would nevertheless be able to use it with the library to build a circuit as if it was an ordinary node:

```python
H0 = tn.Node(h, axis_names=['q0in', 'q0out'])
H1 = tn.Node(h, axis_names=['q1in', 'q1out'])
edge(H0, 'q0out') ^ edge(CZ, 'q0in')
edge(CZ, 'q0out') ^ edge(H1, 'q0in')
CNOT = [H0, CZ, H1]
assert edges(CNOT), {'q0in', 'q1in', 'q0out', 'q1out'}
```

It'd be interesting to see whether these benefits apply in other applications, too.

**Comment by viathor** [comment]
Side story: certain special cases of symmetries in #86 become simpler to address if we treat node collections as first class citizens. For example, a block-diagonal tensor where all blocks are the same submatrix B can be represented as a python 2-tuple `(I, B)` where `I` is the identity matrix. In this example, the node collection that imitates a node is a disconnected graph.

This wouldn't solve #86, but it would take us closer. To actually resolve #86 using node collections, we'd need other tools, such as helper tensors that translate between index set representations and a way to represent addition of networks.

**Comment by amilsted** [comment]
> Functions that work on a Node instance, should work on a collection of Nodes, too.

👍 Like this principle!

Ironic, though, that extending it to the current `n1["axis_name1"] ^ n2["axis_name2"]` syntax (or even the `@` syntax) would actually require a special container class 😄. 
Perhaps there *should* be an explicit `NodeSet`? Then we could do
```python
CNOT = NodeSet([H0, CZ, H1])
CNOT["q0in"] ^ other_gate["q0out"]
```
Then again, I guess the functional edge lookup you propose, `tn.edge(CNOT, "q0in")`, makes the code more readable for the uninitiated.

**Comment by Thenerdstation** [comment]
> It is hard to see how to extend the principle to A @ B

We have 2 options, we could either contract the networks using our `auto` contractor, or we can merge the two node sets into one larger set. I prefer the latter.

**Comment by amilsted** [comment]
> > It is hard to see how to extend the principle to A @ B
> 
> We have 2 options, we could either contract the networks using our `auto` contractor, or we can merge the two node sets into one larger set. I prefer the latter.

Not sure I can agree with that. Would `@` work like this for individual `Node`s too, merging them into a set? Are you suggesting switching to a lazy contraction model (which might be interesting...)? Also, it is already easy to combine sets of nodes: `{n1, n2} | {n3, n4} == {n1, n2, n3, n4}`.

Perhaps node collections should be ordered, with the ordering determining the automatic contraction under `@` and `contract_between()`? In code: `contract_between([n1, n2, n3], n4) == ((n1 @ n2) @ n3) @ n4`. Users can still feed a whole set to a Contractor for optimization if they want. Perhaps this is a bit too subtle though?

Maybe we just shouldn't take the principle this far...?

**Comment by mganahl** [comment]
I too think this is a very interesting thing to try. The biggest issue I can see has already
been discussed above (i.e. how do treat contractions between collections?).

We could only support shortcuts like `a[0] ^ b[0]` or
`a @ b` for elementary `Node` objects (not `Collection[Node]`). 

For the case that the arguments of `contract_between`  are `Node` I think the expected behavior has to be contraction (how else is she dupposed to do contractions otherwise?).

The other cases are less clear. One could argue that we leave the responsibility to the user and simply merge the `Collection`s. However, I feel that supporting `Collection[Node]` becomes pointless at this point, because all we end up doing is accumulating more and more nodes. 
Furthermore, this behavior is essentially already supported by merging of `set` or `list`. So let's examine the contraction option.

I think the guiding principle here should be to assume that the user had a good reason for bunching`Node`s. For this reason I'd rather not silently override this choice by contracting edges within the collection.

Thus,  in these cases I think we should simply `contract` all edges connecting the `Node` and `Collection[Node]`, and leave any `Edge`s within `Collection[Node]` untouched. The reason for using `contract` over `contract_between` is that e.g. for the case that `Node` connects to all nodes in `Collection[Node]`, the latter would end up contracting the edges in the collection as well, violating our promise of not contracting them.

What remains to be decided is whether or not we should use a contraction algorithm.
I'm not sure about this, but we could use a heuristic (e.g. ordering the connecting edges according to their appearance within the first collection) or a contractin finder as default, but also let the user pass a list of edges to specify a desired order.

**Comment by mganahl** [comment]
Hi @MichaelMarien @amilsted @Thenerdstation @viathor, I'd like to raise again the question if and/or how we should proceed with this more seriously. I am essentially finished with removing all dependencies of `TensorNetwork` in the branch `free_node_no_net`. The code there essentially supports everything that google/TensorNetwork supports now on the master branch (apart from the visualization). I've ported all tests that were portable, and everything seems to work fine so far. 
If we want to make a full change, however, we will need to update experiments and examples as well (which I haven't touched). I think right now would be a good time to start this process because the two branches currently still have very similar functionality.

**Comment by Thenerdstation** [comment]
Let's discuss this on our meeting tomorrow. I agree we need to do this sooner rather than later.

**Comment by MichaelMarien** [comment]
Hi @mganahl thanks for all the work. I've been trying to keep up with the posts here, but it's a very busy period at work, so didnt have the time yet to look at your implementations :(.

**Comment by mganahl** [comment]
Hey Chase, can you send me an invite? thanks!

**Comment by Thenerdstation** [comment]
Its the same link from https://github.com/google/TensorNetwork/issues/212

https://meet.google.com/csz-zzin-cme

**Comment by amilsted** [comment]
Just a note on @mganahl's branch: Weak references to `Node`s in `Edge`s mean that any `Node` that falls out of scope in the user's code is implicitly removed from the network. This is probably not what we want! I might want to just keep track of a base node and then use `reachable()` later on, for example. So back to strong refs?

**Comment by mganahl** [comment]
I had them actually removed in an earlier version, but then decided to put them back to leave this decision open to a general discussion. I agree that we probably don't need them anymore (there are a few places in the code where I had to jump through some hoops because of the weakref).

**Comment by amilsted** [comment]
I think the weakref behavior is likely to be *very* counter intuitive for users now there is no network object to keep hold of all the nodes.

**Comment by mganahl** [comment]
Yes, that's totally true, this can produce very obscure behvaiour.  I'm totally in favor of removing them. I think we should start collecting these things. Could you submit an issue to the branch on my fork with these things? Once the transition is done I can move them over to thie repo (otherwise, if no one objects, I'm also fine with opening issues on this repo, might just be a little confusing).

**Comment by Thenerdstation** [comment]
I agree, we should probably remove the weak refs now.

Although, we should be conscience of how Tensors get deallocated outside of a `jit`. We did this weak ref stuff to avoid those nasty OOM errors from before. If we remove the weak refs (which we should), then we need to think about how we deal with these issues

**Comment by MichaelMarien** [comment]
> I'd like to raise again the question if and/or how we should proceed with this more seriously. I am essentially finished with removing all dependencies of `TensorNetwork` in the branch `free_node_no_net`. 

@mganahl , @Thenerdstation any thought on the best way forward? I had some time today and checked the code in the free-node branch (I liked it so far) but I guess some help to migrate this is welcome. Moreover, if we go for it I suggest to close this issue (or limit the scope) and start discussing topics like weak refs in separate issues, that would make it easier to follow and contribute - at least for me.

**Comment by mganahl** [comment]
@MichaelMarien I started the switch in my fork on the remove_tn branch. We decided to split it up into several PRs, and this will be the first one. We‘ll keep TensorNetwork for now, but add the free-node property to all BaseNode. This first PR is essentially only about adding the new functionality without breaking any old code. Once this is in, I suggest we start to remove TensorNetwork bit by bit

**Comment by mganahl** [comment]
> I agree, we should probably remove the weak refs now.
> 
> Although, we should be conscience of how Tensors get deallocated outside of a `jit`. We did this weak ref stuff to avoid those nasty OOM errors from before. If we remove the weak refs (which we should), then we need to think about how we deal with these issues

As long as we keep `TensorNetwork`,  I definitely should be careful about how nodes are garbage collected. That said, I've changed contraction methods to automatically set `node1` and `node2` attributes to `None` once an edge has been contracted (i.e. disabling the edge). I think
this should actually be enough to fix the OOM problems that we were experiencing since we are no longer keeping unused references to `Node`s.

**Comment by Thenerdstation** [comment]
https://github.com/google/TensorNetwork/pull/264 is in now.

I think the only large things that remain are fixing the tests and causing `TensorNetwork` to raise an error correct?

**Comment by mganahl** [comment]
There's still some stuff to do:
1. Change contractors to accept `list` or `set` of `Node` objects instead of a `TensorNetwork` (I have this prepared)
2. remove all `TensorNetwork`s from `ncon` (I'm also working on this one).
3. remove `TensorNetwork` form examples (@amilsted has worked on that and should be able to submit a PR readily I guess)
4. Add convenience methods for saving/loading sets of `Node`s

**Comment by mganahl** [comment]
I'll submit issues for those.

**Comment by Thenerdstation** [comment]
Examples is getting nuked so I'm not worried about that

If you have a PR for the contractors that would be awesome.

@amilsted can do the ncon switch if he has time.

I'll fix the remaining tests. 

This plan sounds good to me! Let's get all of this done asap. I'll get the tests fixed today.

**Comment by Thenerdstation** [comment]
* Correction.
Experiments is getting nuked. Examples will stay.

I'll fix all of examples except for `mera` which requires the contractors getting fixed.

**Comment by mganahl** [comment]
Allright, I'll start working on the contractors. I'll submit a PR for `ncon`  as well since it's already done.

**Comment by amilsted** [comment]
> I'll fix all of examples except for `mera` which requires the contractors getting fixed.

@Thenerdstation I submitted a PR to @mganahl  for this a while ago. He has now noticed this :)

**Comment by mganahl** [comment]
If no one objects I think I will close this issue

**Comment by Thenerdstation** [comment]
Let's wait until your PR is in.

(Actually, have your PR close this use by adding "Fixes #259" to your PR description.

**Comment by Thenerdstation** [comment]
API upgrade is nearly complete. The only thing that is currently not supported/deleted is network saving/loading, which we may just remove.

Next question is what to do for the immediate short term. We have 2 choices.

1. Rip the bandaid off and complete delete `TensorNetwork` for the upcoming release with a very clear error message telling users how they can upgrade their code.
2. Keep `TensorNetwork` for the next 1-2 release cycles and add a warning to users that we will be removing `TensorNetwork` soon.

I prefer 1.

**Comment by mganahl** [comment]
I prefer 2. If we keep  it for  1 cycle, this will give people  some time to adapt to the change. If someone is using the API for some class and has some tight deadline, it will be annoying if suddenly everything is broken and she has to go through the whole code to fix it while some
other stuff might be more pressing. After a month or two there's no more excuse not to upgrade the code, and we can just remove it then .

**Comment by Thenerdstation** [comment]
I guess I see your point. Then I guess all that is left is to upgrade the README and the readthedocs documentation. We can def get this done by the end of the week.

**Comment by mganahl** [comment]
@Thenerdstation I just checked, and I had loading and saving prepared for a PR as well. I need to run some tests and checks on it, but I should be able to open a PR tonight or tomorrow

**Comment by Thenerdstation** [comment]
Great! That's the last thing we needed then.
